File: rb/features.rb
# Joplin SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module JoplinFeatures
  def self.make_feature(name)
    case name
    when "base"
      JoplinBaseFeature.new
    when "debug"
      JoplinDebugFeature.new
    when "idempotency"
      JoplinIdempotencyFeature.new
    when "metrics"
      JoplinMetricsFeature.new
    when "paging"
      JoplinPagingFeature.new
    when "ratelimit"
      JoplinRatelimitFeature.new
    when "retry"
      JoplinRetryFeature.new
    when "test"
      JoplinTestFeature.new
    when "timeout"
      JoplinTimeoutFeature.new
    else
      JoplinBaseFeature.new
    end
  end
end

File: py/joplin_sdk/features.py
# Joplin SDK feature factory

from joplin_sdk.feature.base_feature import JoplinBaseFeature
from joplin_sdk.feature.debug_feature import JoplinDebugFeature
from joplin_sdk.feature.idempotency_feature import JoplinIdempotencyFeature
from joplin_sdk.feature.metrics_feature import JoplinMetricsFeature
from joplin_sdk.feature.paging_feature import JoplinPagingFeature
from joplin_sdk.feature.ratelimit_feature import JoplinRatelimitFeature
from joplin_sdk.feature.retry_feature import JoplinRetryFeature
from joplin_sdk.feature.test_feature import JoplinTestFeature
from joplin_sdk.feature.timeout_feature import JoplinTimeoutFeature


_FEATURES = {
    "base": lambda: JoplinBaseFeature(),
    "debug": lambda: JoplinDebugFeature(),
    "idempotency": lambda: JoplinIdempotencyFeature(),
    "metrics": lambda: JoplinMetricsFeature(),
    "paging": lambda: JoplinPagingFeature(),
    "ratelimit": lambda: JoplinRatelimitFeature(),
    "retry": lambda: JoplinRetryFeature(),
    "test": lambda: JoplinTestFeature(),
    "timeout": lambda: JoplinTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
